runtime/lua/vim/lsp/util.lua

```diff
@@ -171,25 +171,23 @@ end
 ---@param encoding string utf-8|utf-16|utf-32| defaults to utf-16
 ---@return integer byte (utf-8) index of `encoding` index `index` in `line`
 function M._str_byteindex_enc(line, index, encoding)
-  local len = #line
-  if index > len then
-    -- LSP spec: if character > line length, default to the line length.
-    -- https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position
-    return len
-  end
+  local len8 = #line
+  local len32, len16 = vim.str_utfindex(line)
```

Comment:
len32 and len16 are not used in the utf-8 case so this could be moved down to avoid the costs.


```diff
diff --git a/runtime/lua/vim/lsp/util.lua b/runtime/lua/vim/lsp/util.lua
index 1e47fffc40..1652ca9889 100644
--- a/runtime/lua/vim/lsp/util.lua
+++ b/runtime/lua/vim/lsp/util.lua
@@ -170,11 +170,10 @@ end
 ---@param index integer UTF index
 ---@param encoding string utf-8|utf-16|utf-32| defaults to utf-16
 ---@return integer byte (utf-8) index of `encoding` index `index` in `line`
 function M._str_byteindex_enc(line, index, encoding)
   local len8 = #line
-  local len32, len16 = vim.str_utfindex(line)
   -- LSP spec: if character > line length, default to the line length.
   -- https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position
   if not encoding then
     encoding = 'utf-16'
   end
@@ -182,11 +181,13 @@ function M._str_byteindex_enc(line, index, encoding)
     if index and index <= len8 then
       return index
     else
       return len8
     end
-  elseif encoding == 'utf-16' then
+  end
+  local len32, len16 = vim.str_utfindex(line)
+  if encoding == 'utf-16' then
     return index <= len16 and vim.str_byteindex(line, index, true) or len8
   elseif encoding == 'utf-32' then
     return index <= len32 and vim.str_byteindex(line, index) or len8
   else
     error('Invalid encoding: ' .. vim.inspect(encoding))
```